cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_service_kernels_def_dpc.hpp

```diff
@@ -0,0 +1,926 @@
+                                Index(static_cast<Index>(
+                                          data_ptr[id * data_column_count + feat_id]) > feat_bin);
+                            group_row_to_right_count += reduce(sbg, to_right, plus<Index>());
+                        }
+
+                        if (0 == sub_group_local_id) {
+                            group_left_boundary =
+                                atomic_global_add(node_aux + 0,
+                                                  group_row_count - group_row_to_right_count);
+                            group_right_boundary =
+                                atomic_global_add(node_aux + 1, group_row_to_right_count);
+                        }
+                        group_left_boundary = broadcast(sbg, group_left_boundary, 0);
+                        group_right_boundary = broadcast(sbg, group_right_boundary, 0);
+                    }
+
+                    Index group_row_to_right_count = 0;
+                    for (Index i = ind_start + sub_group_local_id; i < ind_end;
+                         i += sub_group_size) {
+                        const Index id = tree_order_ptr[offset + i];
+                        const Index to_right =
+                            Index(static_cast<Index>(data_ptr[id * data_column_count + feat_id]) >
+                                  feat_bin);
+                        const Index boundary =
+                            group_row_to_right_count + exclusive_scan(sbg, to_right, plus<Index>());
+                        const Index pos_new =
+                            (to_right ? left_ch_row_count + group_right_boundary + boundary
+                                      : group_left_boundary + i - ind_start - boundary);
+                        tree_order_buf_ptr[offset + pos_new] = id;
+                        group_row_to_right_count += reduce(sbg, to_right, plus<Index>());
+                    }
+                }
+            }
+        });
+    });
+
+    event.wait_and_throw();
+
+    std::swap(tree_order, tree_order_buf);
+    return event;
+}
+
+template <typename Float, typename Bin, typename Index, typename Task>
+sycl::event train_service_kernels<Float, Bin, Index, Task>::update_mdi_var_importance(
+    const pr::ndarray<Index, 1>& node_list,
+    const pr::ndarray<Float, 1>& node_imp_decrease_list,
+    pr::ndarray<Float, 1>& res_var_imp,
+    Index data_column_count,
+    Index node_count,
+    const be::event_vector& deps) {
+    ONEDAL_ASSERT(node_list.get_count() == node_count * impl_const_t::node_prop_count_);
+    ONEDAL_ASSERT(node_imp_decrease_list.get_count() == node_count);
+    ONEDAL_ASSERT(res_var_imp.get_count() == data_column_count);
+
+    Index local_size = preferable_group_size_;
+    //calculating local size in way to have all subgroups for node in one group to use local buffer
+    while (local_size > node_count && local_size > preferable_sbg_size_) {
+        local_size >>= 1;
+    }
```

Comment:
Looks like there is a function that calculates nxpt power of 2 in primitives common

Reply:
Thanks, replaced it with this func